mixnet/node/src/lib.rs

```diff
@@ -254,100 +246,140 @@ impl PacketForwarder {
         }
+        Ok(self.connections.get_mut(&target).unwrap())
+    }
 
-        let tcp = self.connections.get_mut(&msg.target).unwrap();
-
-        match msg.body.write(tcp).await {
-            Ok(_) => Ok(None),
-            // we should only retry io errors (exclude unsupported for now, may be more in future)
-            Err(ProtocolError::IO(e)) if e.kind() != ErrorKind::Unsupported => {
-                // Update the connection, I actully do not want to do it unless the connection is in broken
-                // situation, but rust does not provide a method to let us check if the connection is broken
-                // or not, so I just hard code the possible situations worth to refresh.
-                if matches!(
-                    e.kind(),
-                    ErrorKind::ConnectionAborted
-                        | ErrorKind::ConnectionReset
-                        | ErrorKind::ConnectionRefused
-                        | ErrorKind::NotConnected
-                        | ErrorKind::BrokenPipe
-                        | ErrorKind::TimedOut
-                ) {
-                    match TcpStream::connect(msg.target).await {
-                        Ok(fresh_tcp) => {
-                            *tcp = fresh_tcp;
-                        }
-                        Err(e) => {
-                            tracing::error!("failed to update connection to {}, the local machine network is down: {e}", msg.target);
-                            return Err(MixnetNodeError::Protocol(ProtocolError::IO(e)));
+    async fn handle_msg(&mut self, msg: Message) {
+        match msg {
+            Message::Packet { target, body } => self.send(target, body, 0).await,
+            Message::Retry {
+                target,
+                body,
+                retry_count,
+            } => self.send(target, body, retry_count).await,
+        }
+    }
+
+    async fn send(&mut self, target: SocketAddr, body: Body, retry_count: usize) {
+        let config = self.config;
+        let tx = self.message_tx.clone();
+        match self.get_connection(target).await {
+            Ok(tcp) => {
+                let retry_delay = config.retry_delay;
+                let max_retries = config.max_retries;
+                if let Err(err) = body.write(tcp).await {
+                    match err {
+                        ProtocolError::IO(e) if e.kind() == std::io::ErrorKind::Unsupported => {
+                            tracing::error!("fail to send message to {target}: {e}");
                         }
+                        _ => Self::handle_retry(
+                            tx,
+                            target,
+                            body,
+                            retry_count,
+                            retry_delay,
+                            max_retries,
+                        ),
                     }
                 }
+            }
+            Err(e) => self.handle_connection_failure(e, target, body, retry_count),
```

Comment:
nit: `handle_connection_failure` and `handle_retry` are quite similar, can we reduce duplication?

One idea could be to have something like but you can get creative!
```
match self.get_connection(target).await.map(|..| body.write(tcp)) {  // async map exists somewhere
    Err(e) if matches!(err.kind(), ...) | Ok(Err(e)) if e.kind() != ErrorKind::Unsupported => {
      // do the retry with a single shared function
   }
   Ok(Ok(_)) => happy case
   // maybe other cases
}
```

Reply:
Also, I don't think we need to make `Message` as enum. I guess we can leave it as it was.